### PR TITLE
feat: remove ChainAccessForFlatStorage

### PR DESCRIPTION
**Pull request by pugachAG**
Part of #8577

This PR includes:
* Store block info as part of `FlatStateDeltaMetadata` on disk. It is used to construct `FlatStorage` instead of `ChainAccessForFlatStorage`.
* Remove `ChainAccessForFlatStorage`.
* Rename `FlatStateDelta` -> `FlatStateChanges` to be consistent with `StateChanges`. We keep the notion of `FlatStateDelta` as changes + metadata.
* Rename `FlatStorage` `add_block` -> `add_delta`
* Remove duplicate `flat::delta` tests from `flat::storage` (that is copy&paste fail from #8664)

Note: this makes proper GC for deltas more critical, so we need to look into #8655 asap.

### CODE REVIEW

File: core/store/src/flat/storage.rs [hunk 1]
```diff
@@ -517,6 +517,10 @@ mod tests {
             *self.height_to_hashes.get(&height).unwrap()
         }
 
+        fn get_block(&self, height: BlockHeight) -> BlockInfo {
```
Comment (Longarithm) [inline core/store/src/flat/storage.rs:520]: Where do we call it?

File: core/store/src/flat/store_helper.rs [hunk 2]
```diff
@@ -36,40 +39,50 @@ impl FlatStateColumn {
...
-    let key = KeyForFlatStateDelta { shard_id, block_hash };
+    let key = KeyForFlatStateDelta { shard_id, block_hash: delta.metadata.block.hash }
+        .try_to_vec()
+        .map_err(|_| FlatStorageError::StorageInternalError)?;
     store_update
-        .set_ser(FlatStateColumn::Deltas.to_db_col(), &key.try_to_vec().unwrap(), delta)
-        .map_err(|_| FlatStorageError::StorageInternalError)
+        .set_ser(FlatStateColumn::Changes.to_db_col(), &key, &delta.changes)
+        .map_err(|_| FlatStorageError::StorageInternalError)?;
+    store_update
```
Comment (Longarithm) [inline core/store/src/flat/store_helper.rs:76]: It's a bit confusing that we set/remove metadata, but never read it. Maybe also implement `get_delta` here and call it on flat storage creation, and get rid of `ChainAccessForFlatStorage` occurrences later?

By this I'm trying to ensure that we write correct metadata even on intermediate branches, so we can safely run flat storage experiments.

Side thought: now it's a bit more concerning that we never read changes from disk on real-time load. Maybe we need to add a test later which limits capacity of cache with changes, or verifies consistency of cache with DB.

File: core/store/src/flat/storage.rs (hunk 1 above)
Comment (pugachAG) [reply core/store/src/flat/storage.rs:520]: in multiple places when creating `FlatStateDelta` for tests

Comment (Longarithm) [reply core/store/src/flat/storage.rs:520]: Oh, I see.

File: core/store/src/flat/store_helper.rs (hunk 2 above)
Comment (pugachAG) [reply core/store/src/flat/store_helper.rs:87]: Updated the code to use metadata for `FlatStorage` creation.

**Review by Longarithm — APPROVED** [review]

File: core/store/src/flat/store_helper.rs (hunk 2 above)
Comment (pugachAG) [reply core/store/src/flat/store_helper.rs:87]: good point, I've added that to the description

**Review by Longarithm — APPROVED** [review]
Reapproving due to significant changes, discussed offline.
Major updates for visibility:
* now no height checks in get_blocks_to_head. but this is okay, because all blocks should be based on head, and triggering check would mean inconsistency of state;
* in add_delta and add_block, we need to make additional check for flat_head, because there is no delta for flat_head.

**Pull request merged**